### ISSUE TITLE
Printing twice in simple test w/ fork & drrun, only reproducible w/ ctest

### ISSUE THREAD

**Issue by hgreving2304** [DynamoRIO/dynamorio#3239]
With attached patch, I am getting

_~/dynamorio/build$ /usr/local/google/home/hgreving/dynamorio/build/suite/tests/bin/linux.epoll_pwait_bug
was here_

With drrun alone

_~/dynamorio/build$ /usr/local/google/home/hgreving/dynamorio/build/suite/tests/bin/linux.epoll_pwait_bug
was here
hgreving@hendrik-desktop:~/dynamorio/build$ /usr/local/google/home/hgreving/dynamorio/build/bin64/drrun -- /usr/local/google/home/hgreving/dynamorio/build/suite/tests/bin/linux.epoll_pwait_bug
<Starting application /usr/local/google/home/hgreving/dynamorio/build/suite/tests/bin/linux.epoll_pwait_bug (144453)>
<Initial options = -no_dynamic_options -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/usr/local/google/home/hgreving/dynamorio/build/lib64/debug/libdynamorio.so' 0x00007fb281470688
>
<(1+x) Handling our fault in a TRY at 0x00007fb2816b7f11>
<spurious rep/repne prefix @0x00007fb28122fe58 (f2 41 ff e3): >
was here
<-- parent 144453 forked child 144454 -->
<Stopping application /usr/local/google/home/hgreving/dynamorio/build/suite/tests/bin/linux.epoll_pwait_bug (144453)>
hgreving@hendrik-desktop:~/dynamorio/build$ <Stopping application /usr/local/google/home/hgreving/dynamorio/build/suite/tests/bin/linux.epoll_pwait_bug (144454)>_

With ctest

_~/dynamorio/build$ ctest -V -I 67,67
UpdateCTestConfiguration  from :/usr/local/google/home/hgreving/dynamorio/build/DartConfiguration.tcl
Parse Config file:/usr/local/google/home/hgreving/dynamorio/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/usr/local/google/home/hgreving/dynamorio/build/DartConfiguration.tcl
Parse Config file:/usr/local/google/home/hgreving/dynamorio/build/DartConfiguration.tcl
Test project /usr/local/google/home/hgreving/dynamorio/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 67
    Start 67: code_api|linux.epoll_pwait_bug

67: Test command: /usr/local/google/home/hgreving/dynamorio/build/bin64/drrun "-s" "90" "-quiet" "-debug" "-killpg" "-stderr_mask" "0xC" "-dumpcore_mask" "0" "-code_api" "--" "/usr/local/google/home/hgreving/dynamorio/build/suite/tests/bin/linux.epoll_pwait_bug"
67: Test timeout computed to be: 1500
67: was here
67: was here_

[..]

See the double "was here". Some process/buffering/race issue? Hasn't been investigated further

[0001-Reproducer-for-double-printf-w-fork.patch.tar.gz](https://github.com/DynamoRIO/dynamorio/files/2512532/0001-Reproducer-for-double-printf-w-fork.patch.tar.gz)

**Comment by egrimley** [comment]
I think this has to do with line buffering. Testing your C program without DynamoRIO:
```
$ ./a.out 
was here
$ ./a.out | cat
was here
was here
$ 
```
One could perhaps modify the test framework to force line buffering, like a terminal, but we don't have any reason for doing that, do we?

**Comment by hgreving2304** [comment]
Indeed, with tools.h/print (that does flush), it doesn't show up. I should have added that to the description above. However, flush can cause multithreaded printfs go out of order and things like that, but we shouldn't see duplicated lines? I think it's worth to look at what's happening here
